### PR TITLE
feat: recurrence for BitVec.mul as repeated shifts for bitblasting

### PR DESCRIPTION
**Pull request by bollu**
This allows @hargoniX to develop bitvec theory in parallel

### CODE REVIEW

File: src/Init/Data/BitVec/Lemmas.lean [hunk 1]
```diff
@@ -726,6 +726,32 @@ theorem getLsb_sshiftRight (x : BitVec w) (s i : Nat) :
...
+  l' = (l' / r') * r' + rem'
+
+/-- Such a remainder always exists. -/
+theorem udiv_euclid_eqn_exists (l r : BitVec w) :
+  ∃ (rem : BitVec w), udivDivisor l r rem := sorry
+
+/-- Such a remainder is unique. -/
+theorem udiv_euclid_eqn_unique (l r rem rem' : BitVec w)
+  (hrem : udivDivisor l r rem) (hrem' : udivDivisor l r rem') :
+  rem = rem' := sorry
```
[the remark this answers is not in the document]
Comment (bollu) [reply src/Init/Data/BitVec/Lemmas.lean]: Sorrys dropped.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 2]
```diff
@@ -1085,6 +1111,247 @@ theorem ofInt_mul {n} (x y : Int) : BitVec.ofInt n (x * y) =
...
+    assert bool(s.check())
+
+    # verify what happens in mulRec for all 's'
+    for nbits_keep in range(1, w):
+        s = Solver()
+        s.add(ForAll(l, ForAll(r, mulRec(l, r, nbits_keep) == ZeroExt(w - nbits_keep - 1, Extract(nbits_keep, 0, l * r)))))
+        print(f"* checking mul eqn for width:'{nbits_keep}': '{s}'.")
+        assert bool(s.check())
+mulExample()
+-/
```
[the remark this answers is not in the document]
Comment (bollu) [reply src/Init/Data/BitVec/Lemmas.lean:1154]: Dropped.

File: src/Init/Data/BitVec/Bitblast.lean [hunk 3]
```diff
@@ -159,6 +159,21 @@ theorem add_eq_adc (w : Nat) (x y : BitVec w) : x + y = (adc x y false).snd := b
 theorem allOnes_sub_eq_not (x : BitVec w) : allOnes w - x = ~~~x := by
   rw [← add_not_self x, BitVec.add_comm, add_sub_cancel]
 
+#check BitVec.ofNat
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Bitblast.lean:162]:
```suggestion
```

File: src/Init/Data/BitVec/Bitblast.lean [hunk 4]
```diff
@@ -235,4 +250,69 @@ theorem sle_eq_carry (x y : BitVec w) :
     x.sle y = !((x.msb == y.msb).xor (carry w y (~~~x) true)) := by
   rw [sle_eq_not_slt, slt_eq_not_carry, beq_comm]
 
+/-! ### mul recurrence for bitblasting -/
+
+/-- Recurrence lemma that  saus that truncating to `i+1` bits and then zero extending to `w`
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Bitblast.lean:255]: > that  saus that 

??

File: src/Init/Data/BitVec/Bitblast.lean (hunk 3 above)
Comment (bollu) [reply src/Init/Data/BitVec/Bitblast.lean]: fixed.

File: src/Init/Data/BitVec/Bitblast.lean (hunk 4 above)
Comment (bollu) [reply src/Init/Data/BitVec/Bitblast.lean]: resolved.

File: src/Init/Data/BitVec/Basic.lean [hunk 5]
```diff
@@ -616,6 +616,13 @@ theorem ofBool_append (msb : Bool) (lsbs : BitVec w) :
 
 end bitwise
 
+section twoPow
+
+/-- `twoPow i` is the bitvector `2^i` if `i < w`, and `0` otherwise. That is, 2 to the power `i`. -/
+def twoPow {w : Nat} (i : Nat) : BitVec w := (1#w) <<< i
+
+end twoPow
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Basic.lean:624]: Why is there a one definition section?

File: src/Init/Data/BitVec/Bitblast.lean [hunk 6]
```diff
@@ -159,6 +159,20 @@ theorem add_eq_adc (w : Nat) (x y : BitVec w) : x + y = (adc x y false).snd := b
 theorem allOnes_sub_eq_not (x : BitVec w) : allOnes w - x = ~~~x := by
   rw [← add_not_self x, BitVec.add_comm, add_sub_cancel]
 
+/-- Adding two bitvectors equals or-ing them if they are 1 in mutually exclusive locations -/
+theorem add_eq_or_of_and_eq_zero {w : Nat} (x y : BitVec w)
+    (h : x &&& y = (0#w)) : x + y = x ||| y := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Bitblast.lean:164]:
```suggestion
    (h : x &&& y = 0#w) : x + y = x ||| y := by
```

does this work?

File: src/Init/Data/BitVec/Lemmas.lean [hunk 7]
```diff
@@ -1085,6 +1132,23 @@ theorem ofInt_mul {n} (x y : Int) : BitVec.ofInt n (x * y) =
   apply eq_of_toInt_eq
   simp
 
+@[simp]
+theorem BitVec.mul_one {x : BitVec w} : x * (1#w) = x := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1136]:
```suggestion
theorem BitVec.mul_one {x : BitVec w} : x * 1#w = x := by
```
Does this not work?

Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean:1136]: Also further below.

File: src/Init/Data/BitVec/Basic.lean (hunk 5 above)
Comment (bollu) [reply src/Init/Data/BitVec/Basic.lean:624]: addressed, moved into bitwise.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 8]
```diff
@@ -1085,6 +1132,23 @@ theorem ofInt_mul {n} (x y : Int) : BitVec.ofInt n (x * y) =
...
+  apply eq_of_toNat_eq
+  simp [toNat_mul, Nat.mod_eq_of_lt x.isLt]
+
+@[simp]
+theorem BitVec.mul_zero {x : BitVec w} : x * (0#w) = (0#w) := by
+  apply eq_of_toNat_eq
+  simp [toNat_mul]
+
+theorem BitVec.mul_add {x y z : BitVec w} :
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:1145]: Can you double-check the theorem name? Maybe it would require a dist_ somewhere?

Comment (bollu) [reply src/Init/Data/BitVec/Lemmas.lean:1145]: It's named after `Nat.mul_add`: 

```lean
protected theorem mul_add (n m k : Nat) : n * (m + k) = n * m + n * k :=
  Nat.left_distrib n m k
```

File: src/Init/Data/BitVec/Lemmas.lean [hunk 9]
```diff
@@ -408,6 +418,29 @@ theorem msb_zeroExtend (x : BitVec w) : (x.zeroExtend v).msb = (decide (0 < v) &
...
+    (BitVec.ofNat v 1).zeroExtend w = BitVec.ofNat w 1 := by
+  ext i
+  obtain ⟨i, hilt⟩ := i
+  simp only [getLsb_zeroExtend, hilt, decide_True, getLsb_ofNat, Bool.true_and,
+    Bool.and_iff_right_iff_imp, decide_eq_true_eq]
+  intros hi1
+  have hv := Nat.testBit_one_eq_true_iff_self_eq_zero.mp hi1
+  omega
+
+
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:443]: Drop one newline here?

File: src/Init/Data/BitVec/Lemmas.lean (hunk 8 above)
Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean:1145]: Actually, would we then need a left_distrib and have the various instantiations?

File: src/Init/Data/BitVec/Bitblast.lean [hunk 10]
```diff
@@ -159,6 +159,20 @@ theorem add_eq_adc (w : Nat) (x y : BitVec w) : x + y = (adc x y false).snd := b
 theorem allOnes_sub_eq_not (x : BitVec w) : allOnes w - x = ~~~x := by
   rw [← add_not_self x, BitVec.add_comm, add_sub_cancel]
 
+/-- Adding two bitvectors equals or-ing them if they are 1 in mutually exclusive locations -/
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Bitblast.lean:162]:
```suggestion
/-- Adding two bitvectors equals or-ing them if they are 1 in mutually exclusive locations. -/
```

File: src/Init/Data/BitVec/Lemmas.lean (hunk 8 above)
Comment (bollu) [reply src/Init/Data/BitVec/Lemmas.lean:1145]: @tobiasgrosser it's unclear to me, I'm trying to develop minimal theory to get the recurrences through, as it should establish a baseline to then improve the status quo. I guess we might want to mirror the `Nat` development --- I'm not sure.

File: src/Init/Data/BitVec/Bitblast.lean [hunk 11]
```diff
@@ -235,4 +249,83 @@ theorem sle_eq_carry (x y : BitVec w) :
...
+        simp [hik', hik'']
+      · have hik'' : ¬ (k < i) := by omega
+        simp [hik', hik'']
+  · ext k
+    simp
+    intros h₁ _ _ _
+    omega
+
+theorem mulRec_eq_mul_signExtend_truncate (l r : BitVec w) (s : Nat) :
+    (mulRec l r s) = l * ((r.truncate (s + 1)).zeroExtend w) := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Bitblast.lean:292]:
```suggestion
    mulRec l r s = l * ((r.truncate (s + 1)).zeroExtend w) := by
```

Does this work?

File: src/Init/Data/BitVec/Lemmas.lean [hunk 12]
```diff
@@ -163,6 +163,16 @@ theorem toNat_zero (n : Nat) : (0#n).toNat = 0 := by trivial
 private theorem lt_two_pow_of_le {x m n : Nat} (lt : x < 2 ^ m) (le : m ≤ n) : x < 2 ^ n :=
   Nat.lt_of_lt_of_le lt (Nat.pow_le_pow_of_le_right (by trivial : 0 < 2) le)
 
+@[simp]
+theorem getLsb_ofBool (b : Bool) (i : Nat) : (BitVec.ofBool b).getLsb i = ((i = 0) && b) := by
```
Comment (tobiasgrosser) [inline src/Init/Data/BitVec/Lemmas.lean:167]:
```suggestion
theorem getLsb_ofBool (b : Bool) (i : Nat) : (BitVec.ofBool b).getLsb i = (i = 0) && b := by
```

does this work?

File: src/Init/Data/BitVec/Lemmas.lean (hunk 8 above)
Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean:1144]: I guess we can leave this as is for now.

**Pull request closed without merging**